eng/versioning/update_versions.py

```diff
@@ -0,0 +1,194 @@
+#    python eng/versioning/update_versions.py --ut all --bt client --tf <pathToPomFile>\pom.xml
+#
+# The script must be run at the root of azure-sdk-for-java.
+
+import argparse
+from datetime import timedelta
+from enum import Enum
+import os
+import re
+import time
+
+version_update_start_marker = re.compile(r'\{x-version-update-start;([^;]+);([^}]+)\}')	
+version_update_end_marker = re.compile(r'\{x-version-update-end\}')
+version_update_marker = re.compile(r'\{x-version-update;([^;]+);([^}]+)\}')
+# regex for the version string is suggested one directly from semver.org
+# it's worth noting that both regular expressions on that page have start
+# of line (^) and end of line ($) anchors which need to be removed since
+# what's being matched is in the middle of the string
+# https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
+version_regex_str = r'(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?'
+
+class UpdateType(Enum):
+    external_dependency = 'external_dependency'
+    library = 'library'
+    all = 'all'
+
+    # defining string is necessary to get ArgumentParser's help output to produce
+    # human readable values of UpdateType
+    def __str__(self):
+        return self.value
+
+class BuildType(Enum):
+    client = 'client'
+    data = 'data'
+    management = 'management'
+    
+    # defining string is necessary to get ArgumentParser's help output to produce
+    # human readable values of BuildType
+    def __str__(self):
+        return self.value
+
+class CodeModule:
+    def __init__(self, module_str):
+        # For library versions there will be up to 3 items resulting from the split
+        # which will be module name, dependency version and current version. For
+        # external dependency versions there should only be 2 items resulting from
+        # the split which will be module name and external dependency version. 
+        items = module_str.split(';')
+        if len(items) == 2: 
+            self.name = items[0]
+            self.dependency = items[1].strip()
+        elif len(items) == 3:
+            self.name = items[0]
+            self.dependency = items[1]
+            self.current = items[2].strip()
+        else: 
+            raise ValueError('unable to parse module string: ' + module_str)
+
+    def __str__(self):
+        # current may or may not exist
+        try:
+            return self.name + ': Dependency version=' + self.dependency + ': Current version=' + self.current
+        except AttributeError:
+            return self.name + ': External Dependency version=' + self.dependency
+
+def update_versions(version_map, target_file):
+
+    newlines = []
+    repl_open, repl_thisline, file_changed = False, False, False
+    print('processing: ' + target_file)
+    try:
+        with open(target_file, encoding='utf-8') as f:
```

Comment:
This function only works well if the tags are well formed. I'd suggest in a follow-up adding some validation that verifies the tags are well formed in the file.

Reply:
I agree, There's room here for someone to shoot themselves in the foot. The plan is to eventually have validation that'll kick during PR submission that checks that everything is tagged correctly (including balanced tags for begin/end tags) but that's a future thing. External dependency tags need to be in place first though.